backupccl: fix EXECUTE privileges

mgartner opened the pull request:
#### backupccl: fix EXECUTE privileges

This commit fixes some bugs with restoring UDFs from a backup:

1. Restoring UDFs from a backup created on a v23.1 cluster will have
   `EXECUTE` privilege added for the public role, matching the new
   default privileges for functions in v23.2. This ensures that the
   backed-up functions can be executed without privilege errors.
2. Restoring a full cluster from a v23.2 backup with UDFs will retain
   the original privileges of the UDF. For example, if EXECUTE
   privileges on the public role have been revoked, they will remain
   revoked after the restore.
3. Restoring UDFs from a partial v23.2 backup, i.e., not a full cluster
   backup, will now have default privileges, including `EXECUTE`
   privilege for the public role.

Fixes #112866

There is no release note because no release includes this issue.

Release note: None

#### upgrade: remove grantExecuteToPublicOnAllFunctions upgrade

The `grantExecuteToPublicOnAllFunctions` upgrade is no longer necessary
because post-serialization changes now grant `EXECUTE` on functions to
the public role. The upgrade has been removed.

Release note: None

mgartner committed:
backupccl: fix EXECUTE privileges

This commit fixes some bugs with restoring UDFs from a backup:

1. Restoring UDFs from a backup created on a v23.1 cluster will have
   `EXECUTE` privilege added for the public role, matching the new
   default privileges for functions in v23.2. This ensures that the
   backed-up functions can be executed without privilege errors.
2. Restoring a full cluster from a v23.2 backup with UDFs will retain
   the original privileges of the UDF. For example, if EXECUTE
   privileges on the public role have been revoked, they will remain
   revoked after the restore.
3. Restoring UDFs from a partial v23.2 backup, i.e., not a full cluster
   backup, will now have default privileges, including `EXECUTE`
   privilege for the public role.

Fixes #112866

There is no release note because no release includes this issue.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114203)
<!-- Reviewable:end -->

dt commented:
`backupccl` test added here looks good in terms of mechanics, however I'll defer to more sql/foundations people on the actual behaviors asserted as desired in the test and all the pkg/sql and catalog changes

mgartner commented on pkg/sql/catalog/catpb/privilege.go:
```diff
@@ -40,6 +40,11 @@ const (
+	// Version23_2 corresponds to descriptors created in 23.2 and onwards. These
+	// descriptors that are function descriptions should have EXECUTE privileges
+	// for the public role.
+	Version23_2
```
Does this require a new cluster version? I'm seeing `TestSystemDatabaseSchemaBootstrapVersionBumped` fail with:

```
Check whether you need to bump systemschema.SystemDatabaseSchemaBootstrapVersion
        and then update prevSystemHash to "7d089e4c69d8d6b164b3d1a04e0f3ad5f487f9138c9585bd00b37f397691a090".
        The current value of SystemDatabaseSchemaBootstrapVersion is 1000023.1-32.
```

mgartner commented on pkg/sql/catalog/bootstrap/bootstrap_test.go:
```diff
@@ -164,7 +164,7 @@ func TestSystemDatabaseSchemaBootstrapVersionBumped(t *testing.T) {
 	// If you need to update this value (i.e. failed this test), check whether
 	// you need to bump systemschema.SystemDatabaseSchemaBootstrapVersion too.
-	const prevSystemHash = "cd85ebe773d840ecaf39fd4ccbba6345f54d7414a95b5bf5be2fa4102c26c650"
+	const prevSystemHash = "7d089e4c69d8d6b164b3d1a04e0f3ad5f487f9138c9585bd00b37f397691a090"
```
Is it ok to bump this without updating `ystemschema.SystemDatabaseSchemaBootstrapVersion` in this case? The system schema should have no UDFs, AFAIK, so I thought just bumping this would be fine.

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114203)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @yuzefovich)

___
*[`pkg/sql/catalog/catpb/privilege.go` line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114203#gh-1389910135:-Nj8N6wTAnUBaHhKJhD7:bgwz0l2) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb22e37249fcd19e31cb1d2655d8751ca2c2d82/pkg/sql/catalog/catpb/privilege.go#L47)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Does this require a new cluster version? I'm seeing `TestSystemDatabaseSchemaBootstrapVersionBumped` fail with:

```
Check whether you need to bump systemschema.SystemDatabaseSchemaBootstrapVersion
        and then update prevSystemHash to "7d089e4c69d8d6b164b3d1a04e0f3ad5f487f9138c9585bd00b37f397691a090".
        The current value of SystemDatabaseSchemaBootstrapVersion is 1000023.1-32.
```
</blockquote></details>

i wouldn't expect this to need a new cluster version. the reason this is happening is bc now the hard-coded system database descriptors that are used if you bootstrap from 23.2 no longer match the descriptors that are created if you upgrade from 23.1 to 23.2. 

the full comment is
```
// TestSystemDatabaseSchemaBootstrapVersionBumped serves as a reminder to bump
// systemschema.SystemDatabaseSchemaBootstrapVersion whenever a new upgrade
// creates or modifies the schema of system tables. We unfortunately cannot  
// programmatically determine if an upgrade should bump the version so by  
// adding a test failure when the initial values change, the programmer and  
// code reviewers are reminded to manually check whether the version should  
// be bumped.
```

i think what we want here is _just_ to change `prevSystemHash`

i also think we might want to use
```
var SystemDatabaseSchemaBootstrapVersion \= clusterversion.ByKey(clusterversion.V23\_2)
```
in pkg/sql/catalog/systemschema/system.go. but i'm asking about that [here](https://cockroachlabs.slack.com/archives/C04N0AS14CT/p1699891327510329)
___
*[`pkg/sql/catalog/funcdesc/func_desc_builder.go` line 121 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114203#-Nj8M6vO3VMl5asJwBIu:-Nj8M6vO3VMl5asJwBIv:b-9x143y) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc6799518ff89cba96d0e7df3bbe670d9d2bc162/pkg/sql/catalog/funcdesc/func_desc_builder.go#L121)):*
> ```Go
> 	// Grant EXECUTE privilege on the function for the public role if the
> 	// descriptor was created before v23.2. This mimics the
> 	// grantExecuteToPublicOnAllFunctions descriptor upgrade.
> ```

an interesting consequence of adding this logic here is that the `grantExecuteToPublicOnAllFunctions` upgrade is no longer necessary. the reason is that there already is an upgrade called `FirstUpgradeFromRelease` that will run all the post-deserialization changes on all descriptors.

i guess there's no harm in keeping `grantExecuteToPublicOnAllFunctions` though.
___
*[`pkg/sql/catalog/funcdesc/func_desc_builder.go` line 129 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114203#-Nj8MVtAEtlUyMHFE-LW:-Nj8MVtAEtlUyMHFE-LX:b-hdi4i1) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc6799518ff89cba96d0e7df3bbe670d9d2bc162/pkg/sql/catalog/funcdesc/func_desc_builder.go#L129)):*
> ```Go
> 			false, /* withGrantOption */
> 		)
> 		desc.GetPrivileges().SetVersion(catpb.Version23_2)
> ```

i believe you need a line like
```
fdb.changes.Add(catalog.GrantedExecuteToPublicRole)
```
(and create the `GrantedExecuteToPublicRole` enum)


<!-- Sent from Reviewable.io -->

mgartner committed:
upgrade: remove grantExecuteToPublicOnAllFunctions upgrade

The `grantExecuteToPublicOnAllFunctions` upgrade is no longer necessary
because post-serialization changes now grant `EXECUTE` on functions to
the public role. The upgrade has been removed.

Release note: None

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114203)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @yuzefovich)

___
*[`pkg/sql/catalog/funcdesc/func_desc_builder.go` line 121 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114203#-Nj8M6vO3VMl5asJwBIu:-Nj8Y2h78a3MvKUoBZNw:b-xwn5kx) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc6799518ff89cba96d0e7df3bbe670d9d2bc162/pkg/sql/catalog/funcdesc/func_desc_builder.go#L121)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

an interesting consequence of adding this logic here is that the `grantExecuteToPublicOnAllFunctions` upgrade is no longer necessary. the reason is that there already is an upgrade called `FirstUpgradeFromRelease` that will run all the post-deserialization changes on all descriptors.

i guess there's no harm in keeping `grantExecuteToPublicOnAllFunctions` though.
</blockquote></details>

I've removed it — the best code is no code.
___
*[`pkg/sql/catalog/funcdesc/func_desc_builder.go` line 129 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114203#-Nj8MVtAEtlUyMHFE-LW:-Nj8bvmS6VL9UnCZJsmF:bqw9btr) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc6799518ff89cba96d0e7df3bbe670d9d2bc162/pkg/sql/catalog/funcdesc/func_desc_builder.go#L129)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i believe you need a line like
```
fdb.changes.Add(catalog.GrantedExecuteToPublicRole)
```
(and create the `GrantedExecuteToPublicRole` enum)
</blockquote></details>

I didn't need it before when this was in `RunRestoreChanges` (or maybe I did?), but I've added it now that I've moved this logic to `RunPostDeserializationChanges`.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114203)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @yuzefovich)

___
*[`pkg/sql/catalog/bootstrap/bootstrap_test.go` line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114203#gh-1389925301:-Nj9B0v-4RXNmRPnxIpd:b-th61xq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7400246af6a42732eeace16d08bb9d9a7e1a956c/pkg/sql/catalog/bootstrap/bootstrap_test.go#L167)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Is it ok to bump this without updating `ystemschema.SystemDatabaseSchemaBootstrapVersion` in this case? The system schema should have no UDFs, AFAIK, so I thought just bumping this would be fine.
</blockquote></details>

yeah i think just bumping the hash should be fine


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm! thanks for taking this on

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114203#-:-NjJ2bFV9f3LooXLUCZm:b5hc2u3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @mgartner)

___
*[`pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions` line 412 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114203#-NjJ2KAoArtU1prpzkWF:-NjJ2KAoArtU1prpzkWG:boatopf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7400246af6a42732eeace16d08bb9d9a7e1a956c/pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions#L412)):*
> ```
> SHOW GRANTS ON FUNCTION add
> ----
> db1 public 100109 add(int8, int8) admin ALL true
> ```

unrelated for now, but i am surprised by our decision to include the function ID in the results, given that SHOW commands are intended for usage by humans, not computers.


<!-- Sent from Reviewable.io -->

mgartner commented:
bors r+

mgartner commented:
*[`pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions` line 412 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114203#-NjJ2KAoArtU1prpzkWF:-NjNm_LD0q6eZJAhqBJM:bco6u3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/7400246af6a42732eeace16d08bb9d9a7e1a956c/pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions#L412)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

unrelated for now, but i am surprised by our decision to include the function ID in the results, given that SHOW commands are intended for usage by humans, not computers.
</blockquote></details>

@chengxiong-ruan any reason we included this? I'll remove it if not.


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
*[`pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions` line 412 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114203#-NjJ2KAoArtU1prpzkWF:-NjNqIN21MlTt5g751vj:b14705g) ([raw file](https://github.com/cockroachdb/cockroach/blob/7400246af6a42732eeace16d08bb9d9a7e1a956c/pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions#L412)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

@chengxiong-ruan any reason we included this? I'll remove it if not.
</blockquote></details>

I don't have a strong reason, just thought it might be good for debugging. But, since we also have function signature, function id is really not needed for sure.


<!-- Sent from Reviewable.io -->

mgartner commented:
I still need to backport this.
